Orange/widgets/unsupervised/owsom.py

```diff
@@ -184,6 +187,96 @@ def disconnected_spin(spin):
 N_ITERATIONS = 200
 
 
+class SomSharedValueCompute:
+    def __init__(self, domain: Domain, model: SOM,
+                 offsets: Union[np.ndarray, None],
+                 scales: Union[np.ndarray, None]):
+        # offsets and scales are made immutable so that they are hashable
+        def immutable(a):
+            if a is not None:
+                a = a.copy()
+                a.flags.writeable = False
+            return a
+
+        self.domain = domain
+        self.model = model
+        self.offsets = immutable(offsets)
+        self.scales = immutable(scales)
+        self.__hash = None
+
+    def __call__(self, data):
+        x = data.transform(self.domain).X
+        cont_x, mask, _, _ = SOM.prepare_data(x, self.offsets, self.scales)
+        winners = np.full((len(data), 2), np.nan)
+        distances = np.full(len(data), np.nan)
+        winners[mask], distances[mask] = self.model.winners(cont_x)
+        winners += 1
+        return winners, distances
+
+    # `offsets` and `scales` are ndarray's (and `model` contains ndarray's)
+    # Unless we __eq__ by identity, we can't properly define hash.
+    def __eq__(self, other):
+        return type(self) is type(other) and \
+               self.domain == other.domain \
+               and self.model is other.model \
+               and np.array_equal(self.offsets, other.offsets) \
+               and np.array_equal(self.scales, other.scales)
+
+    def __hash__(self):
+        if self.__hash is None:
+            self.__hash = hash(
+                (hash(self.domain),
+                 id(self.model),
```

Comment:
Instead of `id(self.model)` I'd just write self.model, and let __hash__ do whatever it does by default - probably something related to ID. :)

Also, no need to make individual hashes of the tuple element, calling just `hash((self.domain, self.model, ...` should suffice.

Reply:
Some time ago I used to have habit of writing `return hash(...) ^ hash(...) ^ hash(...)` until I learned from other's code that this isn't OK. These individual hashes probably came from subconscious. :)